File: php/Python3/PhpLexerBase.py
"""
PHP grammar.
The MIT License (MIT).
Copyright (c) 2015-2017, Ivan Kochurkin (kvanttt@gmail.com), Positive Technologies.
Copyright (c) 2016, Jorrit Kronjee (Python port)
Copyright (c) 2019, Thierry Marianne (thierry.marianne@weaving-the-web.org)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
"""

from antlr4 import *
from antlr4.Token import CommonToken


class PhpLexerBase(Lexer):
    AspTags = True
    _scriptTag = False
    _styleTag = False
    _heredocIdentifier = None
    _prevTokenType = 0
    _htmlNameText = None
    _phpScript = False
    _insideString = False

    def nextToken(self):
        token = super(PhpLexerBase, self).nextToken()

        if token.type == self.PHPEnd or token.type == self.PHPEndSingleLineComment:
            if self._mode == self.SingleLineCommentMode:
                # SingleLineCommentMode for such allowed syntax:
                # // <?php echo "Hello world"; // comment ?>
                self.popMode()
            self.popMode()

            if token.text == "</script>":
                self._phpScript = False
                token.type = self.HtmlScriptClose
            else:
                # Add semicolon to the end of statement if it is absent.
                # For example: <?php echo "Hello world" ?>
                if self._prevTokenType == self.SemiColon or \
                        self._prevTokenType == self.Colon or \
                        self._prevTokenType == self.OpenCurlyBracket or \
                        self._prevTokenType == self.CloseCurlyBracket:
                    token = super(PhpLexerBase, self).nextToken()
                else:
                    token = CommonToken(type=self.SemiColon)
                    token.text = ';'
        elif token.type == self.HtmlName:
            self._htmlNameText = token.text
        elif token.type == self.HtmlDoubleQuoteString:
            if token.text == "php" and self._htmlNameText == "language":
                self._phpScript = True
        elif self._mode == self.HereDoc:
            # Heredoc and Nowdoc syntax support: http://php.net/manual/en/language.types.string.php#language.types.string.syntax.heredoc
            if token.type == self.StartHereDoc or token.type == self.StartNowDoc:
                self._heredocIdentifier = token.text[3:].strip().replace("'", "")
            if token.type == self.HereDocText:
                if self.CheckHeredocEnd(token.text):
                    self.popMode()
                    heredoc_identifier = self.GetHeredocEnd(token.text)
                    if token.text.strip().endswith(';'):
                        text = heredoc_identifier + ";\n"
                        token = CommonToken(type=self.SemiColon)
                        token.text = text
                    else:
                        token = super(PhpLexerBase, self).nextToken()
                        token.text = heredoc_identifier + "\n;"
        elif self._mode == self.PHP:
            if self._channel == self.HIDDEN:
                self._prevTokenType = token.type

        return token

    def GetHeredocEnd(self, text):
        return text.strip().rstrip(';')

    def CheckHeredocEnd(self, text):
        return self.GetHeredocEnd(text) == self._heredocIdentifier

    def IsNewLineOrStart(self, pos):
        return self._input.LA(-1) <= 0 or self._input.LA(-1) == ord('\r') or self._input.LA(-1) == ord('\n')

    def PushModeOnHtmlClose(self):
        self.popMode()
        if self._scriptTag:
            if not self._phpScript:
                self.pushMode(self.SCRIPT)
            else:
                self.pushMode(self.PHP)
            self._scriptTag = False
        elif self._styleTag:
            self.pushMode(self.STYLE)
            self._styleTag = False

    def HasAspTags(self):
        return self.AspTags

    def HasPhpScriptTag(self):
        return self._phpScript

    def PopModeOnCurlyBracketClose(self):
        if self._insideString:
            self._insideString = False
            self._channel = self.SKIP
            self.popMode()

    def ShouldPushHereDocMode(self, pos):
        return self._input.LA(pos) == ord('\r') or self._input.LA(pos) == ord('\n')

    def IsCurlyDollar(self, pos):
        return self._input.LA(pos) == ord('$')

    def SetInsideString(self):
        self._insideString = True

File: php/TypeScript/PhpLexerBase.ts
import {CommonToken, Lexer, Token, CharStream} from "antlr4";
// change the imports below accordingly
import PhpParser from "./PhpParser";
import PhpLexer from "./PhpLexer";

export class PhpLexerBase extends Lexer {
    private AspTags: boolean;
    protected _scriptTag: boolean;
    protected _styleTag: boolean;
    private _heredocIdentifier: string | undefined;
    private _prevTokenType: number;
    private _htmlNameText: string | undefined;
    private _phpScript: boolean;
    private _insideString: boolean;
    protected _mode: number
    protected _channel: number

    static DEFAULT_MODE = 0;
    protected static MORE = -2;
    protected static SKIP = -3;

    protected static DEFAULT_TOKEN_CHANNEL: number = (Token as any).DEFAULT_CHANNEL;
    protected static HIDDEN: number = (Token as any).HIDDEN_CHANNEL;
    protected static MIN_CHAR_VALUE = 0x0000;
    protected static MAX_CHAR_VALUE = 0x10FFFF;

    constructor(input: CharStream) {
        super(input);
        this.AspTags = true;
        this._scriptTag = false;
        this._styleTag = false;
        this._heredocIdentifier = undefined;
        this._prevTokenType = 0;
        this._htmlNameText = undefined;
        this._phpScript = false;
        this._insideString = false;
    }

    nextToken() {
        let token = super.nextToken()

        if (token.type === PhpParser.PHPEnd || token.type === PhpLexer.PHPEndSingleLineComment) {
            if (this._mode === PhpLexer.SingleLineCommentMode) {
                // SingleLineCommentMode for such allowed syntax:
                // // <?php echo "Hello world"; // comment ?>
                this.popMode();
            }
            this.popMode();

            if (token.text === "</script>") {
                this._phpScript = false;
                token.type = PhpLexer.HtmlScriptClose;
            } else {
                // Add semicolon to the end of statement if it is absent.
                // For example: <?php echo "Hello world" ?>
                if (this._prevTokenType === PhpLexer.SemiColon || this._prevTokenType === PhpLexer.Colon || this._prevTokenType === PhpLexer.OpenCurlyBracket || this._prevTokenType === PhpLexer.CloseCurlyBracket) {
                    token = super.nextToken()
                } else {
                    token = new CommonToken(undefined, PhpLexer.SemiColon, undefined, undefined, undefined);
                    token.text = ';';
                }
            }
        }

        else if (token.type === PhpLexer.HtmlName) {
            this._htmlNameText = token.text
        }

        else if (token.type === PhpLexer.HtmlDoubleQuoteString) {
            if (token.text === "php" && this._htmlNameText === "language") {
                this._phpScript = true;
            }
        }

        else if (this._mode === PhpLexer.HereDoc) {
            // Heredoc and Nowdoc syntax support: http://php.net/manual/en/language.types.string.php#language.types.string.syntax.heredoc
            if (token.type === PhpLexer.StartHereDoc || token.type === PhpLexer.StartNowDoc) {
                this._heredocIdentifier = token.text.slice(3).trim().replace(/\'$/, '');
            }

            if (token.type === PhpLexer.HereDocText) {
                if (this.CheckHeredocEnd(token.text)) {
                    this.popMode()
                    const heredocIdentifier = this.GetHeredocEnd(token.text)
                    if (token.text.trim().endsWith(';')) {
                        token = new CommonToken((CommonToken as any).EMPTY_SOURCE, PhpParser.SemiColon,
                            undefined, undefined, undefined)
                        token.text = `${heredocIdentifier};\n`;
                    } else {
                        token = super.nextToken()
                        token.text = `${heredocIdentifier}\n;`;
                    }
                }
            }
        }

        else if (this._mode === PhpLexer.PHP) {
            if (this._channel === PhpLexer.HIDDEN) {
                this._prevTokenType = token.type;
            }
        }

        return token;
    }

    GetHeredocEnd(text: string): string {
        return text.trim().replace(/\;$/, "");
    }

    CheckHeredocEnd(text: string): boolean {
        return this.GetHeredocEnd(text) === this._heredocIdentifier;
    }

    IsNewLineOrStart(pos: number): boolean {
        return this._input.LA(pos) <= 0 || this._input.LA(pos) == '\r'.charCodeAt(0) ||
            this._input.LA(pos) == '\n'.charCodeAt(0)
    }

    PushModeOnHtmlClose() {
        this.popMode();
        if (this._scriptTag) {
            if (!this._phpScript) {
                this.pushMode(PhpLexer.SCRIPT);
            } else {
                this.pushMode(PhpLexer.PHP);
            }
            this._scriptTag = false;
        } else if (this._styleTag) {
            this.pushMode(PhpLexer.STYLE);
            this._styleTag = false;
        }
    }

    HasAspTags(): boolean {
        return this.AspTags;
    }

    HasPhpScriptTag(): boolean {
        return this._phpScript;
    }

    PopModeOnCurlyBracketClose() {
        if (this._insideString) {
            this._insideString = false;
            this.skip;
            this.popMode();
        }
    }

    ShouldPushHereDocMode(pos: number): boolean {
        return this._input.LA(pos) === '\r'.charCodeAt(0) || this._input.LA(pos) === '\n'.charCodeAt(0);
    }

    IsCurlyDollar(pos: number): boolean {
        return this._input.LA(pos) === '$'.charCodeAt(0);
    }

    SetInsideString() {
        this._insideString = true
    }
}
